infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/MachineResolver.java

```diff
@@ -84,12 +95,33 @@ private void normalizeMemory(Container container, InternalMachineConfig machineC
     }
   }
 
+  private Map<String, String> toEnv(CheContainer container) {
+    return container.isMountSources()
+        ? Collections.singletonMap(PROJECTS_MOUNT_PATH_ENV_VAR_NAME, PROJECTS_MOUNT_PATH)
+        : null;
+  }
+
   private Map<String, ? extends org.eclipse.che.api.core.model.workspace.config.Volume>
-      toWorkspaceVolumes(List<Volume> volumes) {
+      toWorkspaceVolumes(CheContainer container) throws InfrastructureException {
 
     Map<String, VolumeImpl> result = new HashMap<>();
 
-    for (Volume volume : volumes) {
+    if (container.isMountSources()) {
+      result.put(PROJECTS_VOLUME_NAME, new VolumeImpl().withPath(PROJECTS_MOUNT_PATH));
+    }
+
+    for (Volume volume : container.getVolumes()) {
```

Comment:
This is not only checking for the error condition but also adds the volumes in the result. Streamed, it would have to look something like:

```java
if (container.getVolumes().stream().peek(v -> results.put(...)).filter(v -> v.getName().equals(...)).findAny().isPresent()) {
  throw ...
}
```

which I think is too much magic..

Reply:
Oh, my bad - didn't see putting to the map